pubnub-core/src/subscription/subscribe_loop_supervisor.rs

```diff
@@ -0,0 +1,161 @@
+            let (channel_tx, channel_rx) = mpsc::channel(10);
+
+            let id_or_retry = if let Some(ref mut control_tx) = self.control_tx {
+                // Send a command to add the channel to the running
+                // subscribe loop.
+
+                debug!("Adding channel {:?} to the running loop", channel);
+
+                let (id_tx, id_rx) = oneshot::channel();
+
+                // TODO: unify interfaces to either use `ListenerType` or
+                // `&str` when we refer to a channel.
+                let listener = ListenerType::Channel(channel.to_string());
+
+                let control_comm_result = control_tx
+                    .send(ControlCommand::Add(listener, channel_tx, id_tx))
+                    .await;
+
+                if control_comm_result.is_err() {
+                    // We got send error, this only happens when the receive
+                    // half of the channel is closed.
+                    // Assuming it was dropped because of being out of
+                    // scope, we conclude the subscribe loop has completed.
+                    // We simply cleanup the control tx, and retry
+                    // subscribing.
+                    // The successive subscribtion attempt will result in
+                    // starting off of a new subscription loop and properly
+                    // registering the channel there.
+                    self.control_tx = None;
+
+                    debug!("Restarting the subscription loop");
+
+                    // This is equivalent to calling the `subscribe` fn
+                    // recursively, given we're in the loop context.
+                    None
+                } else {
+                    // We succesfully submitted the command, wait for
+                    // subscription loop to communicate the subscription ID
+                    // back to us.
+                    let id = id_rx.await.unwrap();
+
+                    // Return the values from the loop.
+                    Some((id, control_tx.clone()))
+                }
+            } else {
+                // Since there's no subscribe loop loop found, spawn a new
+                // one.
+
+                let mut channels = Registry::new();
+                let (id, _) = channels.register(channel.to_string(), channel_tx);
+
+                let (control_tx, control_rx) = mpsc::channel(10);
+                let (ready_tx, ready_rx) = oneshot::channel();
+
+                debug!("Creating the subscribe loop");
+                let subscribe_loop_params = SubscribeLoopParams {
+                    control_rx,
+                    ready_tx: Some(ready_tx),
+                    exit_tx: self.params.exit_tx.clone(),
+
+                    transport: pubnub.transport.clone(),
+
+                    origin: pubnub.origin.clone(),
+                    agent: pubnub.agent.clone(),
+                    subscribe_key: pubnub.subscribe_key.clone(),
```

Comment:
Shouldn't these just be cloned into the `SubscribeLoopSupervisor` struct along with `pubnub.runtime`? I'm worried that the new interface, accepting a reference to a `PubNub` struct, could be unsound if a different struct reference is passed in.

Reply:
Yes, this is one of the issues that unraveled toward the very end of the time frame that I had to work on this, so it's a bit incomplete. In the future, I'd take some time to eliminate extra clones and re-think the ownership model of the entities in a broader sense. I.e., the `origin`, `agent` and `subscribe_key` should most likely belong to the `Transport` implementation itself (maybe not directly, via some common struct for HTTP-based transports) - that way we don't have to even deal with them at this API layer. This would be a part of getting rid of the `url`s in `Transport` API. In fact, I sort of have a plan on how to continue from here on to further improvements. Eliminating references to `PubNub` is one of the major concerns of mine, but there are some blockers that need to be unlocked first before this can be done cleanly.

Reply:
I really like where this is going. It will DRY up a great deal of the data interfaces in here.